File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    private int getNodePos(Node head, Node target) {
        int pos = 0;
        while(head != null) {
            if(head == target)
                return pos;
            else {
                ++pos;
                head = head.next;
            }
        }
        return pos;
    }

    private Node getNodeAtPos(Node head, int pos) {
        while(pos > 0) {
            head = head.next;
            --pos;
        }
        return head;
    }

    public Node copyRandomList(Node head) {
        if(head == null)
            return null;

        Node original_head = head, new_head = new Node(head.val);
        Node curr = new_head;
        
        head = head.next;
        while(head != null) {
            curr.next = new Node(head.val);
            curr = curr.next;
            head = head.next;
        }
        curr = new_head;
        head = original_head;

        while(head != null) {
            if(head.random == null)
                curr.random = null;
            else
                curr.random = getNodeAtPos(new_head, getNodePos(original_head, head.random));
            head = head.next;
            curr = curr.next;
        }

        return new_head;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        curr = head
        copy_head = Node(curr.val) if curr else None
        copy_curr = copy_head
        copy_map = dict()
        while curr:
            copy_map[curr] = copy_curr
            if curr.next:
                copy_curr.next = Node(curr.next.val)
            else:
                copy_curr.next = None
            copy_curr = copy_curr.next
            curr = curr.next
                
        curr = head
        copy_curr = copy_head
        while curr:
            copy_curr.random = copy_map.get(curr.random, None)
            copy_curr = copy_curr.next
            curr = curr.next

        return copy_head